index.rb

```diff
@@ -1,160 +1,250 @@
- #Search "ruby understanding self and main"
- 
- 
- 
- module Enumerable
-  def my_each
-  
-    for i in self 
-       yield(i)
-    end
-
+module Enumerable
+  def my_each(&block)
+    each(&block)
   end
 
   def my_each_with_index
-  
-    for i in self
-      yield(i, self.find_index(i))
-    end
+    return to_enum(:my_each_with_index) unless block_given?
 
+    each do |i|
+      yield(i, find_index(i))
+    end
   end
 
   def my_select
+    return to_enum(:my_select) unless block_given?
 
     my_array = []
 
-    for i in self
-      if yield(i)
-        my_array << i
-      end
+    each do |i|
+      my_array << i if yield(i)
     end
 
-    puts my_array
-
+    my_array
   end
 
-  def my_all?
-    if self.length == 0 then
-      return true
+  # rubocop:disable Metrics/MethodLength
+
+  def my_all?(*args)
+    # Checks for Regexp
+    if args.length == 1 && args.first.instance_of?(Regexp)
+      my_all_boolean = true
+      each do |i|
+        my_all_boolean = false unless args.first.match(i)
+      end
+      return my_all_boolean
+    end
+    # Checks for Class type and for a specific character
+    if args.length == 1
+
+      # If we passed in a class
+      if args.first.is_a?(Class)
+        my_all_boolean = true
+        each do |i|
+          my_all_boolean = false unless i.is_a?(args.first)
+        end
+        return my_all_boolean
+      end
+
+      # If just a specific character is given
+      my_all_boolean = true
+      each do |i|
+        my_all_boolean = false unless i == args.first
+      end
+      return my_all_boolean
     end
-    my_all_boolean = true
-    for i in self
 
-      if !yield(i)
-        my_all_boolean = false
+    # Checks for when a block is not given
+    unless block_given?
+      my_all_boolean = true
+      each do |i|
+        my_all_boolean = false unless i
       end
+      return my_all_boolean
+    end
 
+    # If a block is given
+    my_all_boolean = true
+    each do |i|
+      my_all_boolean = false unless yield(i)
     end
-    puts my_all_boolean
+    my_all_boolean
   end
 
-  def my_any?
-    if self.length == 0 then
-      return false
+  def my_any?(*args)
+    if args.length == 1 && !block_given?
+
+      if args.first.instance_of?(Regexp)
+        my_any_boolean = false
+        each do |i|
+          my_any_boolean = true if args.first.match(i)
+        end
```

Comment:
- [ ] There is also an error in your implementation of the my_any?. `%w[dog door rod blade].any?(Integer)` returns `false` while `%w[dog door rod blade].my_any?(Integer)` return `true`